cpp/include/kvikio/remote_handle.hpp

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+
+#include <aws/core/Aws.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/core/utils/stream/PreallocatedStreamBuf.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+
+#include <kvikio/parallel_operation.hpp>
+#include <kvikio/posix_io.hpp>
+#include <kvikio/utils.hpp>
+
+#ifndef KVIKIO_AWS_SDK_FOUND
+#error "cannot include <kvikio/remote_handle.hpp>, configuration did not find AWS SDK"
+#endif
+
+namespace kvikio {
+namespace detail {
+
+/**
+ * Stream implementation of a fixed size buffer.
+ */
+class BufferAsStream : public Aws::IOStream {
+ public:
+  using Base = Aws::IOStream;
+  explicit BufferAsStream(std::streambuf* buf) : Base(buf) {}
+
+  ~BufferAsStream() override = default;
+};
+
+/**
+ * @brief Given a file path like "s3://<bucket>/<object>", return the name of the bucket and object.
+ *
+ * @param path S3 file path.
+ * @return Pair of strings: [bucket-name, object-name].
+ */
+inline std::pair<std::string, std::string> parse_s3_path(const std::string& path)
+{
+  if (path.empty()) { throw std::invalid_argument("The remote path cannot be an empty string."); }
+  if (path.size() < 5 || path.substr(0, 5) != "s3://") {
+    throw std::invalid_argument("The remote path must start with the S3 scheme (\"s3://\").");
+  }
+  std::string p = path.substr(5);
+  if (p.empty()) { throw std::invalid_argument("The remote path cannot be an empty string."); }
+  size_t pos = p.find_first_of('/');
+  if (pos == 0) { throw std::invalid_argument("The remote path does not contain a bucket name."); }
```

Comment:
OK, so if the string is `s3:///object`, then we raise. But if it's `s3://bucket` we return `(bucket, "")`.

Reply:
Yeah, in this case the server would just response with object not found. 
But I have added an explicit check and added some tests